ledger/complete/wal/compactor.go

```diff
@@ -96,3 +112,25 @@ func (c *Compactor) Run() error {
 	}
 	return nil
 }
+
+func (c *Compactor) cleanupCheckpoints() error {
+	// don't bother listening checkpoints if we keep them all
+	if c.checkpointsToKeep == 0 {
+		return nil
+	}
+	checkpoints, err := c.checkpointer.ListCheckpoints()
+	if err != nil {
+		return fmt.Errorf("cannot list checkpoints: %w", err)
+	}
+	if len(checkpoints) > int(c.checkpointsToKeep) {
```

Comment:
Question:

How does it ensure all the checkpoint files are good (not corrupted) checkpoint files. Does `cleanupCheckpoints` check the file checksum of each checkpoint file?

If we have `001(good), 002(good), 003(bad), 004(bad), 005(bad), 006(bad), 007(bad)`, then we will remove `001(good), 002(good)` the two good checkpoints, and keep all bad ones.

Reply:
No, it doesn't. 
But there a few things to consider here.
1) New way of writing checkpoints should significantly reduce number of failed checkpoints - previously we write directly to a file `checkpoint.0000xxxx` so any interrupted write won't leave the corrupted file as it does now.
2) Removal happens during normal working so new checkpoint should be created eventually
3) We can still recover fully, starting from root checkpoint as usual.
So the risk are significantly reduced, with worst outcome being long(er) startup time.
I'm not sure if those risk are worth adding checkpoint checking for every removal. Thoughts?

Reply:
sounds good